File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String s) {
        int maxLen = 0;
        String res = "";
        
        for(int i=0; i<s.length(); i++){
            int start=i, end = i;
            
            while(start >= 0 && end < s.length() && s.charAt(start) == s.charAt(end)){
               if(maxLen < end-start+1){
                   maxLen = end-start+1;
                   res = s.substring(start, end+1);
               }
               start--;
               end++;
            }
            
            start = i; end = i+1;
            while(start >= 0 && end < s.length() && s.charAt(start) == s.charAt(end)){
               if(maxLen < end-start+1){
                   maxLen = end-start+1;
                   res = s.substring(start, end+1);
               }
               start--;
               end++;
            }
        }
        return res;
    }
}

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n=s.length();
        vector<vector<bool>> dp(n, vector<bool>(n));
        int max_len=1, st=0;
        for(int i=n-1 ; i>=0 ; i--){
            for(int j=i ; j<n ; j++){
                if(i==j){
                    dp[i][j]=1;
                }else if(j-i==1){
                    dp[i][j] = s[i]==s[j];
                }else{
                    dp[i][j] = (s[i]==s[j] && dp[i+1][j-1]);
                }
                if(dp[i][j] && max_len < j-i+1){
                    st=i;
                    max_len = j-i+1;
                }
            }
        }
        return s.substr(st, max_len);
    }
};